lib/mysql/mysql.go

```diff
@@ -0,0 +1,651 @@
+	if terr := conn.SetReadDeadline(time.Now().Add(c.Config.Timeout)); terr != nil {
+		return nil, fmt.Errorf("Error calling SetReadTimeout(): %s", terr)
+	}
+	var header [4]byte
+	n, err := reader.Read(header[:])
+	if err != nil {
+		return nil, fmt.Errorf("Error reading packet header (timeout=%s): %s", err, c.Config.Timeout)
+	}
+	if n != 4 {
+		return nil, fmt.Errorf("Wrong number of bytes returned (got %d, expected 4)", n)
+	}
+	seq := header[3]
+	// length is actually Uint24; clear the bogus MSB before decoding
+	header[3] = 0
+	len := binary.LittleEndian.Uint32(header[:])
+	packet := PacketLogEntry{
+		Length:         len,
+		SequenceNumber: seq}
+
+	// If we fail, we will still have the Length / Sequence Number logged
+	c.pushToPacketLog(&packet)
+	var body = make([]byte, len, len)
+	n, err = reader.Read(body)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading %d bytes: %s", len, err)
+	}
+	// Log the raw body, even if the parsing fails
+	packet.Raw = base64.StdEncoding.EncodeToString(body)
+
+	if seq != c.SequenceNumber {
+		log.Warnf("Sequence number mismatch: got 0x%x, expected 0x%x", seq, c.SequenceNumber+1)
+	}
+	// Update sequence number
+	c.SequenceNumber = seq + 1
+	ret, err := c.decodePacket(body)
+	if err != nil {
+		return nil, fmt.Errorf("Error decoding packet body (length = %d, sequence number = %d): %s", len, seq, err)
+	}
+	packet.Parsed = ret
+
+	return ret, nil
+}
+
+// Perform a TLS handshake using the configured TLSConfig on the current connection
+func (c *Connection) StartTLS() error {
+	client := tls.Client(*c.Connection, c.Config.TLSConfig)
+	err := client.Handshake()
+	if err != nil {
+		return fmt.Errorf("TLS Handshake error: %s", err)
+	}
+	*(c.Connection) = client
+	return nil
+}
+
+// Connect to the configured server and perform the initial handshake
+func (c *Connection) Connect() error {
+	// @TODO @FIXME -- Break this into Connect/Scan/Disconnect?
+	// Allow Scan on pre-connected / user-supplied connections?
+	dialer := net.Dialer{Timeout: c.Config.Timeout}
+	conn, err := dialer.Dial("tcp", fmt.Sprintf("%s:%d", c.Config.Host, c.Config.Port))
+	if err != nil {
+		return fmt.Errorf("Connect error: %s", err)
+	}
+	c.Connection = &conn
+	c.State = STATE_CONNECTED
+
+	packet, err := c.readPacket()
+	if err != nil {
+		return fmt.Errorf("Error reading server handshake packet: %s", err)
+	}
+
+	switch packet.GetType() {
+	case PACKET_TYPE_HANDSHAKE:
+		// OK
+	case PACKET_TYPE_ERROR:
+		return fmt.Errorf("Server returned error after connecting: %s", packet.GetDescription())
+	default:
+		return fmt.Errorf("Server returned unexpected packet type %s after connecting: %s", packet.GetType(), packet.GetDescription())
+	}
+	handshakePacket := packet.(*Handshake_Packet)
+	c.Handshake = handshakePacket
+
+	// How to handle mismatched reserved? It will be available in the output, but should it trigger a 'failure'?
+	// if hex.EncodeToString(parsed.reserved) != "00000000000000000000" { ... }
```

Comment:
I would probably ignore reserved, and/or only output it if it's _not_ `00000....`

Reply:
- [ ] Selectively output